File: tests/engine.test.ts
import { describe, expect, test } from 'bun:test'

import { DEFAULT_SETTINGS, calculate, pickBand } from '../src/lib/engine.ts'
import { DEFAULT_INPUTS } from '../src/lib/defaults.ts'
import type { Inputs, Result, Settings } from '../src/lib/types.ts'

const healthy: Partial<Inputs> = {
  tps_pct_above_19: 100,
  uptime_pct_month: 100,
  concurrent_avg: 5,
  unique_players_week: 50,
  discord_members: 100,
}

const mk = (over: Partial<Inputs>): Inputs => ({ ...DEFAULT_INPUTS, ...healthy, ...over })

const run = (over: Partial<Inputs>, settings: Settings = DEFAULT_SETTINGS): Result =>
  calculate(mk(over), settings)

const ruleIds = (result: Result): string[] => result.alerts.map((a) => a.rule)

const sum = (result: Result): number => {
  const a = result.allocation
  return a.infra + a.ef_fill + a.product + a.growth + a.people + a.infra_buffer + a.unallocated
}

describe('bands', () => {
  test('picks by remaining', () => {
    expect(pickBand(0).id).toBe('micro')
    expect(pickBand(99.99).id).toBe('micro')
    expect(pickBand(100).id).toBe('mid')
    expect(pickBand(499.99).id).toBe('mid')
    expect(pickBand(500).id).toBe('large')
    expect(pickBand(1999.99).id).toBe('large')
    expect(pickBand(2000).id).toBe('mega')
  })

  test('mid band splits 20 / 50 / 30 with no People', () => {
    const r = run({ cash_in_month: 235, infra_cost_month: 35, ef_target_months: 6 })
    expect(r.band?.id).toBe('mid')
    expect(r.remaining).toBe(200)
    expect(r.allocation.ef_fill).toBe(40)
    expect(r.allocation.product).toBe(100)
    expect(r.allocation.growth).toBe(60)
    expect(r.allocation.people).toBe(0)
  })

  test('large band pays People', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35 })
    expect(r.band?.id).toBe('large')
    expect(r.allocation.ef_fill).toBe(200)
    expect(r.allocation.product).toBe(400)
    expect(r.allocation.growth).toBe(250)
    expect(r.allocation.people).toBe(150)
  })

  test('mega band', () => {
    const r = run({ cash_in_month: 3000, infra_cost_month: 1000 })
    expect(r.band?.id).toBe('mega')
    expect(r.allocation.ef_fill).toBe(400)
    expect(r.allocation.product).toBe(700)
    expect(r.allocation.growth).toBe(500)
    expect(r.allocation.people).toBe(400)
  })

  test('micro band sends the leftover to the EF when infra is covered', () => {
    const r = run({ cash_in_month: 85, infra_cost_month: 35 })
    expect(r.band?.id).toBe('micro')
    expect(r.allocation.ef_fill).toBe(50)
    expect(r.allocation.product).toBe(0)
    expect(r.allocation.unallocated).toBe(0)
  })

  test('micro band leaves the leftover as buffer when the EF is full', () => {
    const r = run({ cash_in_month: 85, infra_cost_month: 35, ef_current: 210 })
    expect(r.allocation.ef_fill).toBe(0)
    expect(r.allocation.unallocated).toBe(50)
  })
})

describe('rule 1: infra first', () => {
  test('a shortfall zeroes everything else', () => {
    const r = run({ cash_in_month: 20, infra_cost_month: 35 })
    expect(r.infra_shortfall).toBe(15)
    expect(r.allocation.infra).toBe(20)
    expect(r.allocation.product).toBe(0)
    expect(r.allocation.growth).toBe(0)
    expect(r.allocation.people).toBe(0)
    expect(r.allocation.ef_fill).toBe(0)
    expect(r.band).toBeNull()
    expect(ruleIds(r)).toContain('R1')
  })

  test('exactly covering infra leaves nothing to split', () => {
    const r = run({ cash_in_month: 35, infra_cost_month: 35 })
    expect(r.infra_shortfall).toBe(0)
    expect(r.remaining).toBe(0)
    expect(r.allocation.unallocated).toBe(0)
  })
})

describe('rule 2: EF cap', () => {
  test('cap is infra times ef_target_months', () => {
    const r = run({ cash_in_month: 500, infra_cost_month: 35, ef_target_months: 6 })
    expect(r.ef_cap).toBe(210)
  })

  test('fills only up to the cap and parks the rest', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35, ef_current: 150 })
    expect(r.allocation.ef_fill).toBe(60)
    expect(r.ef_after).toBe(210)
    expect(r.allocation.unallocated).toBe(140)
    expect(ruleIds(r)).toContain('R2')
  })

  test('a full EF sends its whole share to the buffer', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35, ef_current: 210 })
    expect(r.allocation.ef_fill).toBe(0)
    expect(r.allocation.unallocated).toBe(200)
    expect(r.ef_progress_pct).toBe(100)
  })
})

describe('rule 3: stage gates', () => {
  test('TPS below the alpha gate moves Growth to Product', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35, tps_pct_above_19: 94 })
    expect(r.allocation.growth).toBe(0)
    expect(r.allocation.product).toBe(650)
    expect(r.allocation.infra_buffer).toBe(0)
    expect(ruleIds(r)).toContain('R3')
  })

  test('alpha ignores uptime, beta does not', () => {
    const low = { cash_in_month: 1035, infra_cost_month: 35, uptime_pct_month: 99.4 }
    expect(run({ ...low, stage: 'alpha' }).allocation.growth).toBe(250)
    expect(run({ ...low, stage: 'beta', tps_pct_above_19: 99 }).allocation.growth).toBe(250)
    expect(run({ ...low, stage: 'v1', tps_pct_above_19: 99 }).allocation.growth).toBe(0)
  })

  test('unhealthy infra parks the Growth share as infra reserve', () => {
    const r = run({
      cash_in_month: 1035,
      infra_cost_month: 35,
      stage: 'v1',
      tps_pct_above_19: 99,
      uptime_pct_month: 98.5,
    })
    expect(r.allocation.growth).toBe(0)
    expect(r.allocation.infra_buffer).toBe(250)
    expect(r.allocation.product).toBe(400)
  })
})

describe('rule 4: community ratio', () => {
  test('too many Discord members per player kills Growth', () => {
    const r = run({
      cash_in_month: 1035,
      infra_cost_month: 35,
      discord_members: 500,
      unique_players_week: 50,
    })
    expect(r.health.discord_ratio).toBe(10)
    expect(r.allocation.growth).toBe(0)
    expect(r.allocation.product).toBe(650)
    expect(ruleIds(r)).toContain('R4')
  })

  test('the ratio is skipped when there are no players', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35, unique_players_week: 0 })
    expect(r.health.discord_ratio).toBeNull()
    expect(r.allocation.growth).toBe(250)
    expect(ruleIds(r)).toContain('I3')
  })
})

describe('rule 5: load pressure', () => {
  test('high concurrency plus failing TPS parks Growth as infra reserve', () => {
    const r = run({
      cash_in_month: 1035,
      infra_cost_month: 35,
      tps_pct_above_19: 90,
      concurrent_avg: 40,
    })
    expect(r.health.load_pressure).toBe(true)
    expect(r.health.infra_healthy).toBe(false)
    expect(r.allocation.growth).toBe(0)
    expect(r.allocation.infra_buffer).toBe(250)
    expect(ruleIds(r)).toContain('R5')
  })

  test('high concurrency alone changes nothing', () => {
    const r = run({ cash_in_month: 1035, infra_cost_month: 35, concurrent_avg: 40 })
    expect(r.health.load_pressure).toBe(false)
    expect(r.allocation.growth).toBe(250)
  })
})

describe('rule 6: People comes out of profit', () => {
  test('People never exceeds what is left after infra and the EF fill', () => {
    for (const cash of [535, 700, 1035, 2035, 5000]) {
      const r = run({ cash_in_month: cash, infra_cost_month: 35 })
      expect(r.allocation.people).toBeLessThanOrEqual(r.remaining - r.allocation.ef_fill)
    }
  })

  test('no leftover means no People', () => {
    const r = run({ cash_in_month: 35, infra_cost_month: 35 })
    expect(r.allocation.people).toBe(0)
  })
})

describe('runway and totals', () => {
  test('runway counts months of infra covered after this month', () => {
    const r = run({ cash_on_hand_start: 100, cash_in_month: 200, infra_cost_month: 50 })
    expect(r.runway_months).toBe(5)
  })

  test('no infra cost means no runway to report', () => {
    expect(run({ infra_cost_month: 0 }).runway_months).toBeNull()
  })

  test('a short runway is flagged', () => {
    const r = run({ cash_on_hand_start: 0, cash_in_month: 70, infra_cost_month: 35 })
    expect(ruleIds(r)).toContain('I1')
  })

  test('the rows always add up to the cash that came in', () => {
    const cases: Partial<Inputs>[] = [
      { cash_in_month: 0, infra_cost_month: 35 },
      { cash_in_month: 20, infra_cost_month: 35 },
      { cash_in_month: 85.37, infra_cost_month: 35 },
      { cash_in_month: 333.33, infra_cost_month: 35.55, ef_current: 12.5 },
      { cash_in_month: 1234.56, infra_cost_month: 47.89, tps_pct_above_19: 80 },
      { cash_in_month: 9999.99, infra_cost_month: 120.01, ef_current: 700 },
    ]
    for (const c of cases) {
      const r = run(c)
      expect(sum(r)).toBeCloseTo(r.inputs.cash_in_month, 2)
    }
  })
})

File: internal/engine/engine_test.go
package engine

import (
	"math"
	"testing"
)

func healthy(over Inputs) Inputs {
	base := DefaultInputs()
	base.TPSPctAbove19 = 100
	base.UptimePctMonth = 100
	base.ConcurrentAvg = 5
	base.UniquePlayersWeek = 50
	base.DiscordMembers = 100
	return merge(base, over)
}

func merge(base, over Inputs) Inputs {
	if over.Month != "" {
		base.Month = over.Month
	}
	if over.CashInMonth != 0 {
		base.CashInMonth = over.CashInMonth
	}
	if over.CashOnHandStart != 0 {
		base.CashOnHandStart = over.CashOnHandStart
	}
	if over.InfraCostMonth != 0 {
		base.InfraCostMonth = over.InfraCostMonth
	}
	if over.EFCurrent != 0 {
		base.EFCurrent = over.EFCurrent
	}
	if over.EFTargetMonths != 0 {
		base.EFTargetMonths = over.EFTargetMonths
	}
	if over.TPSPctAbove19 != 0 {
		base.TPSPctAbove19 = over.TPSPctAbove19
	}
	if over.UptimePctMonth != 0 {
		base.UptimePctMonth = over.UptimePctMonth
	}
	if over.DiscordMembers != 0 {
		base.DiscordMembers = over.DiscordMembers
	}
	if over.DiscordNetGrowthMonth != 0 {
		base.DiscordNetGrowthMonth = over.DiscordNetGrowthMonth
	}
	if over.UniquePlayersWeek != 0 {
		base.UniquePlayersWeek = over.UniquePlayersWeek
	}
	if over.ConcurrentAvg != 0 {
		base.ConcurrentAvg = over.ConcurrentAvg
	}
	if over.Stage != "" {
		base.Stage = over.Stage
	}
	if over.Notes != "" {
		base.Notes = over.Notes
	}
	return base
}

func run(t *testing.T, over Inputs) Result {
	t.Helper()
	return Calculate(healthy(over), NovumPreset())
}

func hasRule(r Result, id string) bool {
	for _, a := range r.Alerts {
		if a.Rule == id {
			return true
		}
	}
	return false
}

func TestPickBand(t *testing.T) {
	p := NovumPreset()
	cases := []struct {
		n  float64
		id string
	}{
		{0, "micro"}, {99.99, "micro"}, {100, "mid"}, {499.99, "mid"},
		{500, "large"}, {1999.99, "large"}, {2000, "mega"},
	}
	for _, c := range cases {
		if got := PickBand(p, c.n).ID; got != c.id {
			t.Fatalf("pickBand(%v)=%s want %s", c.n, got, c.id)
		}
	}
}

func TestMidBand(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 235, InfraCostMonth: 35, EFTargetMonths: 6})
	if r.Band == nil || r.Band.ID != "mid" {
		t.Fatalf("band %+v", r.Band)
	}
	if r.Remaining != 200 {
		t.Fatalf("remaining %v", r.Remaining)
	}
	if r.Allocation.EFFill != 40 || r.Allocation.Product != 100 || r.Allocation.Growth != 60 || r.Allocation.People != 0 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestLargeBandPaysPeople(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 1035, InfraCostMonth: 35})
	if r.Band.ID != "large" {
		t.Fatalf("band %s", r.Band.ID)
	}
	if r.Allocation.EFFill != 200 || r.Allocation.Product != 400 || r.Allocation.Growth != 250 || r.Allocation.People != 150 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestMegaBand(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 3000, InfraCostMonth: 1000})
	if r.Band.ID != "mega" {
		t.Fatalf("band %s", r.Band.ID)
	}
	if r.Allocation.EFFill != 400 || r.Allocation.Product != 700 || r.Allocation.Growth != 500 || r.Allocation.People != 400 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestMicroBandToEF(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 85, InfraCostMonth: 35})
	if r.Band.ID != "micro" {
		t.Fatalf("band %s", r.Band.ID)
	}
	if r.Allocation.EFFill != 50 || r.Allocation.Product != 0 || r.Allocation.Unallocated != 0 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestMicroBandEFFull(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 85, InfraCostMonth: 35, EFCurrent: 210})
	if r.Allocation.EFFill != 0 || r.Allocation.Unallocated != 50 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestR1Shortfall(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 20, InfraCostMonth: 35})
	if r.InfraShortfall != 15 || r.Allocation.Infra != 20 || r.Band != nil {
		t.Fatalf("shortfall result %+v alloc %+v", r.InfraShortfall, r.Allocation)
	}
	if !hasRule(r, "R1") {
		t.Fatal("missing R1")
	}
}

func TestExactlyCoveringInfra(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 35, InfraCostMonth: 35})
	if r.InfraShortfall != 0 || r.Remaining != 0 || r.Allocation.Unallocated != 0 {
		t.Fatalf("got remaining=%v unalloc=%v short=%v", r.Remaining, r.Allocation.Unallocated, r.InfraShortfall)
	}
}

func TestEFCapValue(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 500, InfraCostMonth: 35, EFTargetMonths: 6})
	if r.EFCap != 210 {
		t.Fatalf("ef_cap %v", r.EFCap)
	}
}

func TestEFCapOverflow(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 1035, InfraCostMonth: 35, EFCurrent: 150})
	if r.Allocation.EFFill != 60 || r.EFAfter != 210 || r.Allocation.Unallocated != 140 {
		t.Fatalf("alloc %+v after %v", r.Allocation, r.EFAfter)
	}
	if !hasRule(r, "R2") {
		t.Fatal("missing R2")
	}
}

func TestFullEFToBuffer(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 1035, InfraCostMonth: 35, EFCurrent: 210})
	if r.Allocation.EFFill != 0 || r.Allocation.Unallocated != 200 || r.EFProgressPct != 100 {
		t.Fatalf("alloc %+v pct %v", r.Allocation, r.EFProgressPct)
	}
}

func TestR3TPSToProduct(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 1035, InfraCostMonth: 35, TPSPctAbove19: 94})
	if r.Allocation.Growth != 0 || r.Allocation.Product != 650 || r.Allocation.InfraBuffer != 0 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
	if !hasRule(r, "R3") {
		t.Fatal("missing R3")
	}
}

func TestAlphaIgnoresUptimeBetaDoesNot(t *testing.T) {
	low := Inputs{CashInMonth: 1035, InfraCostMonth: 35, UptimePctMonth: 99.4}
	alpha := run(t, merge(low, Inputs{Stage: "alpha"}))
	if alpha.Allocation.Growth != 250 {
		t.Fatalf("alpha growth %v", alpha.Allocation.Growth)
	}
	beta := run(t, merge(low, Inputs{Stage: "beta", TPSPctAbove19: 99}))
	if beta.Allocation.Growth != 250 {
		t.Fatalf("beta growth %v", beta.Allocation.Growth)
	}
	v1 := run(t, merge(low, Inputs{Stage: "v1", TPSPctAbove19: 99}))
	if v1.Allocation.Growth != 0 {
		t.Fatalf("v1 growth %v", v1.Allocation.Growth)
	}
}

func TestUnhealthyInfraParksGrowth(t *testing.T) {
	r := run(t, Inputs{
		CashInMonth: 1035, InfraCostMonth: 35, Stage: "v1",
		TPSPctAbove19: 99, UptimePctMonth: 98.5,
	})
	if r.Allocation.Growth != 0 || r.Allocation.InfraBuffer != 250 || r.Allocation.Product != 400 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
}

func TestR4CommunityRatio(t *testing.T) {
	r := run(t, Inputs{
		CashInMonth: 1035, InfraCostMonth: 35,
		DiscordMembers: 500, UniquePlayersWeek: 50,
	})
	if r.Health.DiscordRatio == nil || *r.Health.DiscordRatio != 10 {
		t.Fatalf("ratio %+v", r.Health.DiscordRatio)
	}
	if r.Allocation.Growth != 0 || r.Allocation.Product != 650 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
	if !hasRule(r, "R4") {
		t.Fatal("missing R4")
	}
}

func TestRatioSkippedWithoutPlayers(t *testing.T) {
	in := healthy(Inputs{CashInMonth: 1035, InfraCostMonth: 35})
	in.UniquePlayersWeek = 0
	r := Calculate(in, NovumPreset())
	if r.Health.DiscordRatio != nil {
		t.Fatalf("ratio %+v", r.Health.DiscordRatio)
	}
	if r.Allocation.Growth != 250 {
		t.Fatalf("growth %v", r.Allocation.Growth)
	}
	if !hasRule(r, "I3") {
		t.Fatal("missing I3")
	}
}

func TestR5LoadPressure(t *testing.T) {
	r := run(t, Inputs{
		CashInMonth: 1035, InfraCostMonth: 35,
		TPSPctAbove19: 90, ConcurrentAvg: 40,
	})
	if !r.Health.LoadPressure || r.Health.InfraHealthy {
		t.Fatalf("health %+v", r.Health)
	}
	if r.Allocation.Growth != 0 || r.Allocation.InfraBuffer != 250 {
		t.Fatalf("alloc %+v", r.Allocation)
	}
	if !hasRule(r, "R5") {
		t.Fatal("missing R5")
	}
}

func TestHighConcurrencyAlone(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 1035, InfraCostMonth: 35, ConcurrentAvg: 40})
	if r.Health.LoadPressure || r.Allocation.Growth != 250 {
		t.Fatalf("health %+v alloc %+v", r.Health, r.Allocation)
	}
}

func TestPeopleFromProfit(t *testing.T) {
	for _, cash := range []float64{535, 700, 1035, 2035, 5000} {
		r := run(t, Inputs{CashInMonth: cash, InfraCostMonth: 35})
		if r.Allocation.People > r.Remaining-r.Allocation.EFFill {
			t.Fatalf("people %v exceeds profit for cash %v", r.Allocation.People, cash)
		}
	}
}

func TestNoLeftoverNoPeople(t *testing.T) {
	r := run(t, Inputs{CashInMonth: 35, InfraCostMonth: 35})
	if r.Allocation.People != 0 {
		t.Fatalf("people %v", r.Allocation.People)
	}
}

func TestRunway(t *testing.T) {
	r := run(t, Inputs{CashOnHandStart: 100, CashInMonth: 200, InfraCostMonth: 50})
	if r.RunwayMonths == nil || *r.RunwayMonths != 5 {
		t.Fatalf("runway %+v", r.RunwayMonths)
	}
}

func TestNoInfraNoRunway(t *testing.T) {
	in := healthy(Inputs{})
	in.InfraCostMonth = 0
	r := Calculate(in, NovumPreset())
	if r.RunwayMonths != nil {
		t.Fatalf("runway %+v", r.RunwayMonths)
	}
}

func TestShortRunwayFlagged(t *testing.T) {
	r := run(t, Inputs{CashOnHandStart: 0, CashInMonth: 70, InfraCostMonth: 35})
	if !hasRule(r, "I1") {
		t.Fatal("missing I1")
	}
}

func TestRowsAddUp(t *testing.T) {
	cases := []Inputs{
		{CashInMonth: 0, InfraCostMonth: 35},
		{CashInMonth: 20, InfraCostMonth: 35},
		{CashInMonth: 85.37, InfraCostMonth: 35},
		{CashInMonth: 333.33, InfraCostMonth: 35.55, EFCurrent: 12.5},
		{CashInMonth: 1234.56, InfraCostMonth: 47.89, TPSPctAbove19: 80},
		{CashInMonth: 9999.99, InfraCostMonth: 120.01, EFCurrent: 700},
	}
	for _, c := range cases {
		r := run(t, c)
		if math.Abs(r.Allocation.Sum()-r.Inputs.CashInMonth) > 0.005 {
			t.Fatalf("sum %v != cash %v alloc %+v", r.Allocation.Sum(), r.Inputs.CashInMonth, r.Allocation)
		}
	}
}

func TestGenericPolicyDifferentShares(t *testing.T) {
	p := NovumPreset()
	p.Name = "Diplomatica"
	p.Bands = []Band{
		{ID: "only", Label: "todo", Min: 0, Shares: map[string]float64{"ef": 0, "product": 1, "growth": 0, "people": 0}},
	}
	in := healthy(Inputs{CashInMonth: 135, InfraCostMonth: 35})
	r := Calculate(in, p)
	if r.Allocation.Product != 100 || r.Allocation.Growth != 0 || r.Allocation.EFFill != 0 {
		t.Fatalf("generic alloc %+v", r.Allocation)
	}
	if math.Abs(r.Allocation.Sum()-135) > 0.005 {
		t.Fatalf("sum %v", r.Allocation.Sum())
	}
}
